### PR TITLE
Information about SQLite limitations

### LINKED ISSUE #64664: Grafana 9.4.3 dies silently (on high HDD load?)
**What happened**:

I updated Grafana to 9.4.3 (from 9.3.8). Everything looks good but from time to time (and at least every night) Grafana is just not working anymore. I think this happens when the HDD load is high.

**What you expected to happen**:

Grafana should keep running

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

I set the loglevel to debug, then started updating my FreeBSD ports. Short time after that Grafana was not running anymore. The last lines in the Logfile are

```
logger=alertmanager org=1 t=2023-03-13T07:15:09.327144881+01:00 level=debug component=alertmanager orgID=1 component=dispatcher msg="Received alert" alert=CO2-Alarm[24842ab][active]
logger=datasources rule_uid=lp_in6Rnk org_id=1 t=2023-03-13T07:15:09.689849835+01:00 level=debug msg="Querying for data source via SQL store" uid=L5jNJVZ7z orgId=1
logger=ngalert.state.historian rule_uid=TaUBLwd4z org_id=1 t=2023-03-13T07:15:09.991378496+01:00 level=debug msg="Done saving alert annotation batch"
logger=ngalert.state.historian rule_uid=EtFi3-x4z org_id=1 t=2023-03-13T07:15:10.901337212+01:00 level=debug msg="Done saving alert annotation batch"
logger=datasources t=2023-03-13T07:15:14.007294232+01:00 level=error msg="Failed getting data source" err="database is locked" uid=cb_4ryF7k id=0 name= orgId=1
logger=sqlstore.transactions rule_uid=aYormIFVz org_id=1 t=2023-03-13T07:15:14.007424053+01:00 level=info msg="Database locked, sleeping then retrying" error="database is locked" retry=1 code="database is locked"
```

**Environment**:
- Grafana version: 9.4.3
- Data source type & version: InfluxDB 1.8
- OS Grafana is installed on: FreeBSD 13.1
- User OS & Browser: 
- Grafana plugins:
- Others:

### PR DESCRIPTION
**Pull request by usmangt**
I am creating this PR to add more content about the SQLite (the default database for Grafana) limitations. SQLite is fine for small environments but not suitable when monitoring many Dashboards and sending lots of alerts frequently.

For e.g. this issue is just from this week and then got followed up by 3-4 more similar issues (all related to SQLite and are linked with this issue)

Closes #64664 

By adding more information, will have less frequency of such issues reported by users.

Also, once this PR get submitted, will also create a new PR about the possible workaround to not get the error message "database locked" while using SQLite.

**What is this feature?**

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by mbrancato** [comment]
@usmangt I think it would be good to describe I/O requirements. Realistically, Grafana needs very little storage.  In cloud deployments such as on Kubernetes, small PVCs = small IOPS / disk bandwidth. It does seem that without WAL mode, Grafana is not compatible with slow disks, even in moderate alerting usage. 

I did go through my history, the SQLite limitations didn't happen till after Grafana 9.x where we were forced onto the new alerting system.

It appears as if under the default operation, the SQLite database allows writers to block readers, and this appears to return an error. I think that Grafana should consider enabling WAL mode by default, given that without it, there is no MVCC implementation - writers are going to block readers and general concurrency issues will occur. It's a 12+ year old feature, so I think its probably fairly stable.
https://www.sqlite.org/wal.html

See my other issue on the wrong / poorly descriptive error type being returned: #64692

**Comment by usmangt** [comment]
Hi @mbrancato 

Thanks for the feedback.

My main objective of this PR is to provide some more information when using SQLite so that users who not do know in advance which backend Database they want to choose accordingly before installing Grafana and scaling it out and started to run into issues (if their systems are in production environments). 

For e g.especially in [HA environments](https://grafana.com/docs/grafana/latest/setup-grafana/set-up-for-high-availability/) where you have to switch to either MySQL or PostgreSQL. (I will add this part in this PR so as to make it more visible).


Another example is that if someone is using containerized environments such as Kubernetes etc. and get such an error (or wants HA) then have to go all the way to migrate their database from SQLite to either MySQL or PostgreSQL (which is off-course a downtime for them and need to tackle it before proceeding forward).


> I think that Grafana should consider enabling WAL mode by default

Yes, that could be considered and I also wanted to anyway write this part in our troubleshooting guide. But I will create a request for our Engineering team and they will review it. :+1: 

> See my other issue on the wrong / poorly descriptive error type being returned: https://github.com/grafana/grafana/issues/64692

I am already working on that one :slightly_smiling_face: and saw your new reply (will follow up there)

**Comment by mbrancato** [comment]
@usmangt I think that regardless of how many alerts are setup, this is an error users could see. I increased the IOPS on my disk by 5x and I still saw this error (only once so far). If there is any concurrency in Grafana (its using Go, so likely yes here), you're always going to be working in a race condition between multiple access attempts to the database.

What you're saying here is it is just "less likely" - the race condition still exists as Grafana uses concurrency but the DB has no concurrency control mechanism. Instead of documenting this, I think its important to understand the underlying issue is not documentation, but implementation. That might require that Grafana itself use concurrency control mechanisms like a shared semaphore or mutex to control access to the DB to avoid concurrency when there is no MVCC.

https://pkg.go.dev/sync#Mutex

**Review by chri2547 — CHANGES_REQUESTED** [review]
Thanks for the contribution! Please take a look at my suggestion and see how that works for you. When you are ready for me to approve, please ask me to re-review.

File: docs/sources/setup-grafana/installation/_index.md [hunk 1]
```diff
@@ -54,6 +54,8 @@ Grafana supports the following databases:
 
 By default, Grafana installs with and uses SQLite, which is an embedded database stored in the Grafana installation location.
 
+> **Note:** SQLite works well if your environment is small, but is not recommended when your environment starts growing. For more information about the limitations of SQLite, refer to [Appropriate Uses For SQLite](https://www.sqlite.org/whentouse.html). If you want high availability, you must use either a MySQL or PostgreSQL database. For information about how to define the database configuration parameters inside the `grafana.ini` file, refer to [[database]](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#database).
```
Comment (chri2547) [inline docs/sources/setup-grafana/installation/_index.md:57]:
```suggestion
> **Note:** SQLite works well if your environment is small, but is not recommended when your environment starts growing. For more information about the limitations of SQLite, refer to [Appropriate Uses For SQLite](https://www.sqlite.org/whentouse.html). If you want high availability, you must use a MySQL or PostgreSQL database. For information about how to define the database configuration parameters inside the `grafana.ini` file, refer to [[database]](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#database).
```

**Review by chri2547 — APPROVED** [review]
Looks great! Thank you for the contribution!

**Comment by GeorgeMaged** [comment]
Hello there, I am using Grafana 9.3.6 with the default sqlite3 DB, And after creating alarms that are monitoring a large number of network interfaces if they exceed the threshold or not, Since then Grafana started crashing and I found many logs having the error " database is locked" . I tried when I read here in the community enabling the WAL but It didn't make such a difference. Do I need to consider migrating into mysql or postgres?

**Pull request merged**